src/ucp/proto/proto_am.inl

```diff
@@ -38,41 +38,67 @@ ucs_status_t ucp_do_am_bcopy_multi(uct_pending_req_t *self, uint8_t am_id_first,
+
+    for (;;) {
+        if (offset == 0) {
+            /* First */
+            packed_len = uct_ep_am_bcopy(uct_ep, am_id_first, pack_first, req, 0);
+            UCS_PROFILE_REQUEST_EVENT_CHECK_STATUS(req, "am_bcopy_first", packed_len,
+                                                   packed_len);
+        } else if (offset + max_middle < req->send.length) {
+            /* Middle */
+            packed_len = uct_ep_am_bcopy(uct_ep, am_id_middle, pack_middle, req, 0);
+            ucs_assertv((packed_len < 0) || (packed_len <= max_middle + hdr_size_middle),
+                        "packed_len=%zd max_middle=%zu hdr_size_middle=%zu",
+                        packed_len, max_middle, hdr_size_middle);
+            UCS_PROFILE_REQUEST_EVENT_CHECK_STATUS(req, "am_bcopy_middle",
+                                                   packed_len, packed_len);
+        } else {
+            /* Last */
+            packed_len = uct_ep_am_bcopy(uct_ep, am_id_last, pack_last, req, 0);
+            UCS_PROFILE_REQUEST_EVENT_CHECK_STATUS(req, "am_bcopy_last", packed_len,
+                                                   packed_len);
+            if (packed_len >= 0) {
+                return UCS_OK;
+            }
+        }
 
-    if (offset == 0) {
-        /* First */
-        packed_len = uct_ep_am_bcopy(uct_ep, am_id_first, pack_first, req, 0);
-        UCS_PROFILE_REQUEST_EVENT_CHECK_STATUS(req, "am_bcopy_first", packed_len,
-                                               packed_len);
-    } else if (offset + max_middle < req->send.length) {
-        /* Middle */
-        packed_len = uct_ep_am_bcopy(uct_ep, am_id_middle, pack_middle, req, 0);
-        ucs_assertv((packed_len < 0) || (packed_len <= max_middle + hdr_size_middle),
-                    "packed_len=%zd max_middle=%zu hdr_size_middle=%zu",
-                    packed_len, max_middle, hdr_size_middle);
-        UCS_PROFILE_REQUEST_EVENT_CHECK_STATUS(req, "am_bcopy_middle",
-                                               packed_len, packed_len);
-    } else {
-        /* Last */
-        packed_len = uct_ep_am_bcopy(uct_ep, am_id_last, pack_last, req, 0);
-        UCS_PROFILE_REQUEST_EVENT_CHECK_STATUS(req, "am_bcopy_last", packed_len,
-                                               packed_len);
-        return (packed_len < 0) ? packed_len : UCS_OK;
+        if (packed_len < 0) {
+            if (req->send.lane != req->send.pend_add_lane) {
+                // got NO_RES on a different lane than pending one
+                // resched
+                pending_adde_res = ucp_request_pending_add(req, &status);
+                if (!pending_adde_res) {
+                    /* failed to switch req to pending queue, try again */
+                    continue;
+                }
+                ucs_assert(pending_adde_res == 1);
+                ucs_assert(status == UCS_INPROGRESS);
+                return UCS_ERR_PENDING; // remove from current pending
```

Comment:
pls remove commend, and update UCS_ERR_PENDING to use
#define UCP_ERR_PENDING (UCS_ERR_LAST - 1)

Reply:
done